account/watcher/watcher.go

```diff
@@ -0,0 +1,103 @@
+package watcher
+
+import (
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec"
+)
+
+// expiryWatcher implements the ExpiryWatcher interface
+type expiryWatcher struct {
+	handlers EventHandler
+
+	// bestHeight is the height we believe the current chain is at.
+	bestHeight uint32
+
+	// expirations keeps track of the current accounts we're
+	// watching expirations for.
+	expirations map[[33]byte]uint32
+
+	// expirationsPerHeight keeps track of all registered accounts
+	// that expire at a certain height.
+	expirationsPerHeight map[uint32][]*btcec.PublicKey
+
+	expirationsMtx sync.Mutex
+}
+
+// NewExpiryWatcher instantiates a new ExpiryWatcher.
+func NewExpiryWatcher(handlers EventHandler) *expiryWatcher { // nolint:golint
+	return &expiryWatcher{
+		handlers:             handlers,
+		expirations:          make(map[[33]byte]uint32),
+		expirationsPerHeight: make(map[uint32][]*btcec.PublicKey),
+	}
+}
+
+// NewBlock updates the current bestHeight.
+func (w *expiryWatcher) NewBlock(bestHeight uint32) {
+	w.bestHeight = bestHeight
+}
+
+// OverdueExpirations handles the expirations for the given block.
+func (w *expiryWatcher) OverdueExpirations(blockHeight uint32) {
```

Comment:
I think a better design could be to have a priority queue of accounts where the priority is defined by the block height. This would allow us to expire everything before and at the specified block height.

Reply:
Or alternatively make newblock/expiry atomic.